common/src/main/java/io/pravega/common/util/AbstractBufferView.java

```diff
@@ -0,0 +1,200 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.hash.HashHelper;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import lombok.Getter;
+import lombok.SneakyThrows;
+
+/**
+ * Base implementation of {@link BufferView}, providing common functionality.
+ * Derived classes may override these methods with more efficient implementations tailored to their data types.
+ */
+public abstract class AbstractBufferView implements BufferView {
+    static final BufferView EMPTY = new EmptyBufferView();
+    private static final HashHelper HASH = HashHelper.seededWith(AbstractBufferView.class.getName());
+
+    @Override
+    public int hashCode() {
+        HashHelper.HashBuilder builder = HASH.newBuilder();
+        collect(builder::put);
+        return builder.getAsInt();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj instanceof BufferView) {
+            return equals((BufferView) obj);
+        }
+        return false;
+    }
+
+    /**
+     * Checks for equality against another {@link BufferView} instance.
+     *
+     * Note: this method provides a general implementation of equality. Derived classes should override this method with
+     * more efficient implementations if available.
+     *
+     * @param other The other {@link BufferView} instance.
+     * @return True if this instance and the other instance have the same contents.
+     */
+    @SneakyThrows(IOException.class)
+    public boolean equals(BufferView other) {
+        int l = getLength();
+        if (l != other.getLength()) {
+            return false;
+        }
+
+        if (l > 0) {
+            BufferView.Reader thisReader = getBufferViewReader();
+            BufferView.Reader otherReader = other.getBufferViewReader();
+            for (int i = 0; i < l; i++) {
+                if (thisReader.readByte() != otherReader.readByte()) {
```

Comment:
There are multiple implementations of `ByteBuffer`:
- `ArrayView` based (`ByteArraySegment`)
- `CompositeArrayView` (sparse array allocation)
- `CompositeBufferView`
- `ByteBufWrapper`.

I need a general solution that works for all of them. Then each class can have specialized implementations for this. For example, `ByteArraySegment` (by far the most commonly used class) has a special path for comparing against arrays.

PS: `Arrays.equals` only compares whole arrays; I cannot compare array subsegments. In addition, it does nothing special; it just iterates through the two arrays and compares offset-by-offset.

Reply:
Arrays.equals is indeed special. It is calling ArraysSupport.vectorizedMismatch which in the base implementation is comparing 8 bytes at once but is annotated @HotSpotIntrinsicCandidate and so may be optimized further at runtime. It is OK to have a fallback, but if at all possible we should be calling into these accelerated methods in `ByteBuffer` or `Arrays`.

Reply:
Perhaps there is another way to make this more efficient: Rather than having the impls try to specialize this we could have a protected abstract method for this class that returned: `List<ByteBuffer` which has the same bytes as what is inside the buffer view. Then that can be used to implement: (equals, hashcode, readInt, etc) in the generic logic here. I am not aware of anything which can't return that sort of view, but it's OK if we want to create an impl that can't, it just has to not extend the abstract class and implement the interface itself.

Reply:
According to both my IDE's decompilation and the JDK code [here](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/util/Arrays.java#l2680), there is nothing special done by `Arrays.equals(...)`. What Java version are you looking at?

`BufferView` returns both `List<ByteBuffer>` and `collect(Consumer<ByteBuffer>)` (the latter is more efficient as no lists are allocated. In order to make this work, though, I'd have to make this method handle the fact that component buffers may not be of the same length (even though their composition is equal). That would mean a lot of slicing and invoking of `ByteBuffer.equals`.

However if I look at the `ByteBuffer.equals()` method it does nothing special. It simply loops through all the bytes in the buffers and compares byte-by-byte. There is no optimization employed and neither `DirectByteBuffer` nor `HeapByteBuffer` override that. I do not think we'd be gaining much from using that.

Reply:
Resolving for now. We can take advantage of these faster methods once we upgrade to java 11.